scripts/setup-gcloud-cluster.sh

```diff
@@ -125,26 +129,118 @@ gcloud_create_or_update_managed_instance_group  \
+for staking_addr in "${!staking_addrs_dict[@]}" ; do
+  yarn hardhat stake --network hardhat --amount 1000000000000000000000 \
+    --privatekey "${staking_addrs_dict[staking_addr]}"
+done
+
+# Get names of all instances in this cluster
+declare instance_names
+instance_names=$(gcloud_get_managed_instance_group_instances_names "${cluster_id}")
+declare instance_names_arr=( ${instance_names} )
+
+# Prepare sorted staking account addresses so we ensure a stable order of assignment
+declare staking_addresses_arr=( "${!staking_addrs_dict[@]}" )
+readarray -t staking_addresses_arr < <(for addr in "${!staking_addrs_dict[@]}"; do echo "$addr"; done | sort)
+
+# These arrays will hold IP addresses, peer IDs and staking addresses
+# for instance VMs in the encounter order of the `instance_names` array
+declare -a ip_addrs
+declare -a hopr_addrs
+declare -a used_staking_addrs
+
+# Iterate through all VM instances
+# The loop should be parallelized in future to accommodate better with larger clusters
+for instance_idx in "${!instance_names_arr[@]}" ; do
+  # Firstly, retrieve the IP address of this VM instance
+  instance_name="${instance_names_arr[instance_idx]}"
+  node_ip=$(gcloud_get_ip "${instance_name}")
+
+  # All VM instances in the deployed cluster will get a special INFO tag
+  # which contains all handy information about the HOPR instance running in the VM.
+  # These currently include: node wallet address, node peer ID, associated staking account
+  # These information are constant during the lifetime of the VM and
+  # do not change during re-deployment.
+  info_tag=$(gcloud_get_node_info_tag "${instance_name}")
+
+  # Values contained in the INFO tag
+  declare wallet_addr
+  declare peer_id
+  declare staking_addr
+  if [[ -z "${info_tag}" ]]; then
+    # If the instance does not have the INFO tag yet, we need to retrieve all info
+    wallet_addr=$(get_native_address "${api_token}@${node_ip}:3001")
+    peer_id=$(get_hopr_address "${api_token}@${node_ip}:3001")
+
+    # NOTE: We leave only the first public node unstaked
+    if [[ ${instance_idx} -eq 0 && "${docker_image}" != *-nat:* ]]; then
+      staking_addr="unstaked"
+    else
+      # Staking accounts are assigned round-robin
+      staking_addr_idx=$(( (instance_idx ) % ${#staking_addresses_arr[@]} ))
+      staking_addr="${staking_addresses_arr[staking_addr_idx]}"
+    fi
+
+    # Save the info tag
+    info_tag="info:native_addr=${wallet_addr};peer_id=${peer_id};nr_staking_addr=${staking_addr}"
+    gcloud_add_instance_tags "${instance_name}" "${info_tag}"
+  else
+    # Retrieve all information from the INFO tag
+    wallet_addr=$(echo "${info_tag}" | sed -E 's/.*native_addr=([Xxa-f0-9A-F]+).*/\1/g')
+    peer_id=$(echo "${info_tag}" | sed -E 's/.*peer_id=([a-zA-Z0-9]+).*/\1/g')
+    staking_addr=$(echo "${info_tag}" | sed -E 's/.*nr_staking_addr=([a-zA-Z0-9]+).*/\1/g')
+  fi
+
+  ip_addrs+=( "${node_ip}" )
```

Comment:
Doesn't this also adds the IP address of NATed nodes?

Reply:
Technically it does not need to for the NAT nodes, because the array of IPs is used only for awaiting the p2p port to come up in public nodes (which is not possible for NATed). But that should not hurt anything.

Reply:
I rather meant that it also adds NATed nodes to the ip_addrs array which is later on used to wait for port listening. Which, in case of a NATed node, could take forever.

Reply:
The condition on L242 handles this. The script runs first for public and than for NAT nodes.